### ISSUE TITLE
torch.svd seems to have accuracy problem

### ISSUE THREAD

**Issue by wangg12** [pytorch/pytorch#18558]
I have a simple IRLS algorithm implemented with pytorch here [https://github.com/wangg12/IRLS_tensorflow2/blob/master/src/IRLS_pytorch.py](https://github.com/wangg12/IRLS_tensorflow2/blob/master/src/IRLS_pytorch.py), which used `torch.svd` or `torch.pinverse`.

However, the accuracy is much lower than the same algorithm implemented in tensorflow [https://github.com/wangg12/IRLS_tensorflow2/blob/master/src/IRLS_tf_v2.py](https://github.com/wangg12/IRLS_tensorflow2/blob/master/src/IRLS_tf_v2.py) using almost the same APIs.

Similar observation has been reported [https://discuss.pytorch.org/t/torch-pinverse-seems-to-be-inaccurate/33616](https://discuss.pytorch.org/t/torch-pinverse-seems-to-be-inaccurate/33616), [https://github.com/pytorch/pytorch/issues/10965](https://github.com/pytorch/pytorch/issues/10965), and [https://github.com/pytorch/pytorch/issues/16076](https://github.com/pytorch/pytorch/issues/16076).

Is this inaccute behaviour expected?

**Comment by SsnL** [comment]
Hi, could you save the problem matrix in a checkpoint and attach it here? Also, are you running on CPU or GPU? Could you use the issue template to let us know important information such as magma version? It would also be very helpful if you can post the svd outputs in tf. Thanks!

**Comment by wangg12** [comment]
I am running on the GPU.
This is my environment info:
```
Collecting environment information...
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 410.78
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.14.6
[pip3] numpydoc==0.8.0
[pip3] torch==1.0.0
[pip3] torchvision==0.2.1
[conda] mkl                       2018.0.3                      1    
[conda] mkl-include               2018.0.3                      1    
[conda] pytorch                   1.0.0           py3.6_cuda10.0.130_cudnn7.4.1_1  [cuda100] 
[conda] torchvision               0.2.1                     <pip>
```

The accuracies of algorithm above: pytorch ~0.66, tf ~0.85.

And the corresponding matrix and tf svd outputs are [here](https://github.com/wangg12/IRLS_tensorflow2/tree/master/src).

**Comment by SsnL** [comment]
You have a bug at https://github.com/wangg12/IRLS_tensorflow2/blob/721675bb9dd0ba19223e10b346aa639a159a58da/src/IRLS_pytorch.py#L111

`y` has shape `torch.Size([32561, 1])` yet `y_pred` has shape `torch.Size([32561])`.

For the record, you don't need to add explicit `expand` to broadcast everywhere, pytorch does automatic broadcasting, just like numpy.

I fixed that, and use the same `pinv` method as you did in your `tf` script, removed every discrepancy between pt and tf scripts (e.g., l2 reg), and got the same accuracy.

Here is the script: https://gist.github.com/SsnL/acd7ae2e096b01daab0fc7621dd47bc3

Here is the output:

```
Random Seed:  2196
start training...
L2 param(lambda): 20
iter: 0
         neg log likelihood: 23909.580078125
         train acc: 0.24080955982208252, test acc: 0.23622629046440125
         L2 norm of w: 0.11135528236627579
iter: 1
         neg log likelihood: 12469.365234375
         train acc: 0.8444765210151672, test acc: 0.8457711935043335
         L2 norm of w: 2.1622300148010254
         diff of w_old and w: 2.1692447662353516
iter: 2
         neg log likelihood: 11109.1064453125
         train acc: 0.8469641804695129, test acc: 0.8494564294815063
         L2 norm of w: 3.1962523460388184
         diff of w_old and w: 1.1507797241210938
iter: 3
         neg log likelihood: 10792.7138671875
         train acc: 0.8475169539451599, test acc: 0.8501321077346802
         L2 norm of w: 3.856876850128174
         diff of w_old and w: 0.8543247580528259
iter: 4
         neg log likelihood: 10746.015625
         train acc: 0.8479469418525696, test acc: 0.8507463335990906
         L2 norm of w: 4.167870044708252
         diff of w_old and w: 0.49175599217414856
iter: 5
         neg log likelihood: 10743.853515625
         train acc: 0.8479162454605103, test acc: 0.8506848812103271
         L2 norm of w: 4.240683078765869
         diff of w_old and w: 0.1431959867477417
iter: 6
         neg log likelihood: 10743.8466796875
         train acc: 0.8479162454605103, test acc: 0.8507463335990906
         L2 norm of w: 4.244931221008301
         diff of w_old and w: 0.01007823646068573
iter: 7
         neg log likelihood: 10743.84375
         train acc: 0.8479162454605103, test acc: 0.8507463335990906
         L2 norm of w: 4.244948387145996
         diff of w_old and w: 4.568263830151409e-05
training done.

```

**Comment by SsnL** [comment]
TL; DR: pytorch's svd is not the root cause.

**Comment by SsnL** [comment]
@wangg12 I created this PR https://github.com/wangg12/IRLS_tensorflow2/pull/2

**Comment by wangg12** [comment]
@SsnL Thanks for figuring out the bug. The pytorch version was implemented two years ago when the version was about 0.12. At that time, I had to explicitly broadcast everywhere. 

Happy to know that pytorch has no precision issue compared to tf at least in the `svd` operation.

**Comment by wangg12** [comment]
I have update my pseudo inverse algorithm according to your suggestion, which could match `torch.pinverse` and performs slightly better especially when L2 param is 0 at least in my problem.
